presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaFilterManager.java

```diff
@@ -0,0 +1,319 @@
+                KafkaColumnHandle columnHandle = (KafkaColumnHandle) entry.getKey();
+                if (!columnHandle.isInternal()) {
+                    break;
+                }
+                switch (columnHandle.getName()) {
+                    case PARTITION_OFFSET_FIELD:
+                        offsetRanged = getRange(entry.getValue());
+                        break;
+                    case PARTITION_ID_FIELD:
+                        partitionIdsFiltered = getSingleSet(entry.getValue(), partitionIds);
+                        break;
+                    case OFFSET_TIMESTAMP_FIELD:
+                        offsetTimestampRanged = getRange(entry.getValue());
+                        break;
+                    default:
+                        break;
+                }
+            }
+
+            // push down offset
+            if (offsetRanged.isPresent()) {
+                Range range = offsetRanged.get();
+                partitionBeginOffsets = calcTopicPartitionBeginOffsetMap(partitionBeginOffsets,
+                        p -> (range.getBegin() >= 0) ? Optional.of(range.getBegin()) : Optional.empty());
+                partitionEndOffsets = calcTopicPartitionEndOffsetMap(partitionEndOffsets,
+                        p -> (range.getEnd() >= 0) ? Optional.of(range.getEnd()) : Optional.empty());
+            }
+
+            // push down timestamp if possible
+            if (offsetTimestampRanged.isPresent()) {
+                try (KafkaConsumer<byte[], byte[]> kafkaConsumer = consumerFactory.create()) {
+                    Optional<Range> finalOffsetTimestampRanged = offsetTimestampRanged;
+                    partitionBeginOffsets = calcTopicPartitionBeginOffsetMap(partitionBeginOffsets,
+                            p -> findOffsetsByTimestamp(kafkaConsumer, p, finalOffsetTimestampRanged.get().getBegin()));
+                    if (isTimestampUpperBoundPushdownEnabled(session, kafkaTableHandle.getTopicName())) {
+                        partitionEndOffsets = calcTopicPartitionEndOffsetMap(partitionEndOffsets,
+                                p -> findOffsetsByTimestamp(kafkaConsumer, p, finalOffsetTimestampRanged.get().getEnd()));
+                    }
+                }
+            }
+
+            // push down partitions
+            final Set<Long> finalPartitionIdsFiltered = partitionIdsFiltered;
+            List<PartitionInfo> partitionFilteredInfos = partitionInfos.stream()
+                    .filter(partitionInfo -> finalPartitionIdsFiltered.contains((long) partitionInfo.partition()))
+                    .collect(toImmutableList());
+            return new KafkaFilteringResult(partitionFilteredInfos, partitionBeginOffsets, partitionEndOffsets);
+        }
+        return new KafkaFilteringResult(partitionInfos, partitionBeginOffsets, partitionEndOffsets);
+    }
+
+    private boolean isTimestampUpperBoundPushdownEnabled(ConnectorSession session, String topic)
```

Comment:
Add comment describing logic and motivation here. The phrasing I suggested for commit message will do.